li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/producer/LiKafkaFederatedProducerImpl.java

```diff
@@ -0,0 +1,274 @@
+    this(new LiKafkaProducerConfig(props), keySerializer, valueSerializer, largeMessageSegmentSerializer, mdsClient);
+  }
+
+  public LiKafkaFederatedProducerImpl(Map<String, ?> configs) {
+    this(new LiKafkaProducerConfig(configs), null, null, null, null);
+  }
+
+  public LiKafkaFederatedProducerImpl(Map<String, ?> configs,
+      Serializer<K> keySerializer,
+      Serializer<V> valueSerializer,
+      Serializer<LargeMessageSegment> largeMessageSegmentSerializer,
+      MetadataServiceClient mdsClient) {
+    this(new LiKafkaProducerConfig(configs), keySerializer, valueSerializer, largeMessageSegmentSerializer, mdsClient);
+  }
+
+  @SuppressWarnings("unchecked")
+  private LiKafkaFederatedProducerImpl(LiKafkaProducerConfig configs,
+      Serializer<K> keySerializer,
+      Serializer<V> valueSerializer,
+      Serializer<LargeMessageSegment> largeMessageSegmentSerializer,
+      MetadataServiceClient mdsClient) {
+    _clusterGroup = new ClusterGroupDescriptor(configs.getString(LiKafkaProducerConfig.CLUSTER_ENVIRONMENT_CONFIG),
+        configs.getString(LiKafkaProducerConfig.CLUSTER_GROUP_CONFIG));
+
+    // Each per-cluster producer and auditor will be intantiated when the client starts producing to that cluster.
+    _producers = new HashMap<String, PerClusterProducerContext<K, V>>();
+
+    try {
+      // Instantiate metadata service client if necessary.
+        _mdsClient = mdsClient != null ? mdsClient :
+            configs.getConfiguredInstance(LiKafkaProducerConfig.METADATA_SERVICE_CLIENT_CLASS_CONFIG,
+                MetadataServiceClient.class);
+
+      // Register this federated client with the metadata service. The metadata service will assign a UUID to this
+      // client, which will be used for later interaction between the metadata service and the client.
+      //
+      // Registration may also return further information such as the metadata server version and any protocol settings.
+      // We assume that such information will be kept and used by the metadata service client itself.
+      _clientId = mdsClient.registerFederatedClient(_clusterGroup, configs.originals());
+
+      // Instantiate the key serializer if necessary.
+      _keySerializer = keySerializer != null ? keySerializer
+          : configs.getConfiguredInstance(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, Serializer.class);
+      _keySerializer.configure(configs.originals(), true);
+      // Instantiate the key serializer if necessary.
+      _valueSerializer = valueSerializer != null ? valueSerializer
+          : configs.getConfiguredInstance(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, Serializer.class);
+      _valueSerializer.configure(configs.originals(), false);
+
+      // prepare to handle large messages.
+      _largeMessageEnabled = configs.getBoolean(LiKafkaProducerConfig.LARGE_MESSAGE_ENABLED_CONFIG);
+      _maxMessageSegmentSize = configs.getInt(LiKafkaProducerConfig.MAX_MESSAGE_SEGMENT_BYTES_CONFIG);
```

Comment:
this is fragile - the actual limit is the max request size the broker will accept. is there a way to get this from kafka ?

Reply:
This is directly copied from the existing implementation. Not sure why we haven't bound it.

There is a producer-side config called max.request.size, which limits the size of a single request. Also, there is a separate broker-side config called max.message.bytes, which defines the largest record batch size. I assume one should set max.request.size = max.message.bytes.

I can rewrite this to min(max.message.segment.bytes, max.request.size). What do you think?

Reply:
better than nothing, i guess

Reply:
Updated